docs/history-service.md

```diff
@@ -0,0 +1,274 @@
+
+- Similarly, for a workflow task: [`processWorkflowTask`](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/transfer_queue_active_task_executor.go#L217) results in an [RPC to Matching Service](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/transfer_queue_task_executor_base.go#L152) to create the Workflow Task.
+</details><br>
+
+#### Timer Task queue
+
+There are several scenarios in which it's necessary to set a timer associated with a workflow execution and do something when the timer expires. For example:
+
+- A Timer Task representing a `workflow.sleep()` call in user workflow code is processed, when the timer expires, by creating a Transfer Task specifying that a Workflow Task should be enqueued in Matching to advance the workflow.
+- A Timer Task is created on receiving a `ScheduleActivity` command, representing the [schedule-to-start](https://docs.temporal.io/activities#schedule-to-start-timeout) timeout.
+
+These are implemented as tasks in the Timer Task queue.
+A Timer Task becomes available to be executed at the trigger time that is stored with the task.
+
+<details>
+<summary><i>Code entrypoints</i></summary>
+
+- For the Timer Task queue, [`Execute`](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/timer_queue_active_task_executor.go#L90) handles the task differently according to the type of timer.
+
+- For example, [`executeUserTimerTimeoutTask`](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/timer_queue_active_task_executor.go#L136) handles a `workflow.sleep()` timer set in user code by appending `TimerFired`, and `WorkflowTaskScheduled` events to Workflow History, and creating a Transfer Task for the Workflow Task, using the same [state transition machinery](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/workflow/context.go#L565) as we have seen elsewhere.
+
+- Other types of timer task are handled similarly (e.g. `ActivityTimeout`, `WorkflowTaskTimeout`). In contrast, [`executeActivityRetryTimerTask`](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/timer_queue_active_task_executor.go#L426) calls Matching directly, since activity retries do not contribute events to Workflow History.
+
+</details><br>
+
+### State transitions
+
+Notice that a workflow execution "state transition" can occur in response to three different types of event:
+
+- RPC from User Application
+- RPC from Worker
+- Timer fired
+
```

Comment:
what about start child/signal external from another workflow? you could say they're indirectly from an rpc from a worker (they do come from workflow task commands), but it's indirect, the rpc comes from another history shard's transfer queue. so maybe that's a fourth type.

Reply:
Thanks, added

> - A different Workflow Execution performing start-child-workflow / signal-external-workflow